pkg/spec/resources_test.go

```diff
@@ -237,4 +211,120 @@ func TestCreateServices(t *testing.T) {
 	}
 }
 
+func TestParsePortMapping(t *testing.T) {
+	tests := []struct {
+		name        string
+		portMapping string
+		servicePort *api_v1.ServicePort
+		success     bool
+	}{
+		{
+			name:        "Nothing is passed, not even port",
+			portMapping: "",
+			servicePort: nil,
+			success:     false,
+		},
+		{
+			name:        "Only 'port' is passed",
+			portMapping: "1337",
+			servicePort: &api_v1.ServicePort{
+				Port: 1337,
+				TargetPort: intstr.IntOrString{
+					IntVal: 1337,
+				},
+				Protocol: api_v1.ProtocolTCP,
+			},
+			success: true,
+		},
+		{
+			name:        "port:targetPort is passed",
+			portMapping: "1337:1338",
+			servicePort: &api_v1.ServicePort{
+				Port: 1337,
+				TargetPort: intstr.IntOrString{
+					IntVal: 1338,
+				},
+				Protocol: api_v1.ProtocolTCP,
+			},
+			success: true,
+		},
+		{
+			name:        "port/protocol is passed",
+			portMapping: "1337/UDP",
+			servicePort: &api_v1.ServicePort{
+				Port: 1337,
+				TargetPort: intstr.IntOrString{
+					IntVal: 1337,
+				},
+				Protocol: api_v1.ProtocolUDP,
+			},
+			success: true,
+		},
+		{
+			name:        "port:targetPort/protocol is passed",
+			portMapping: "1337:1338/UDP",
+			servicePort: &api_v1.ServicePort{
+				Port: 1337,
+				TargetPort: intstr.IntOrString{
+					IntVal: 1338,
+				},
+				Protocol: api_v1.ProtocolUDP,
+			},
+			success: true,
+		},
+		{
+			name:        "Invalid protocol (neither TCP nor UDP) is passed",
+			portMapping: "1337:1338/INVALID",
+			servicePort: nil,
+			success:     false,
+		},
+		{
+			name:        "Multiple protocols passed, multiple '/' test",
+			portMapping: "1337/TCP:1338/TCP",
+			servicePort: nil,
+			success:     false,
+		},
+		{
+			name:        "Non int port is passed",
+			portMapping: "batman:1338/TCP",
+			servicePort: nil,
+			success:     false,
+		},
+		{
+			name:        "Non int targetPort is passed",
+			portMapping: "1337:batman/TCP",
+			servicePort: nil,
+			success:     false,
+		},
+		{
+			name:        "More than 2 ports passed, multiple ':' test",
+			portMapping: "1337:1338:1339/TCP",
+			servicePort: nil,
+			success:     false,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+
+			sp, err := parsePortMapping(test.portMapping)
+
+			switch test.success {
+			case true:
+				if err != nil {
+					t.Errorf("Expected test to pass but got an error -\n%v", err)
+				}
+			case false:
+				if err == nil {
+					t.Errorf("Expected %v to fail, but test passed!", test.portMapping)
+				}
+			}
+
+			if !reflect.DeepEqual(sp, test.servicePort) {
+				t.Errorf("Expected ServicePort to be -\n%v\nBut got -\n%v", spew.Sprint(test.servicePort), spew.Sprint(sp))
```

Comment:
Forked the discussion from https://github.com/kedgeproject/kedge/pull/242#discussion_r138333477

Consider a case if the test is expected to pass, but it fails then the test should not go any further than 

```go
t.Errorf("Expected test to pass but got an error -\n%v", err)
```
which is line #3315

because we are sure that it will surely throw error at line #324 which is 

```go
t.Errorf("Expected ServicePort to be -\n%v\nBut got -\n%v", spew.Sprint(test.servicePort), spew.Sprint(sp))
```

because the variable `sp` is `nil` because of failed test. So if we know further in path if a test is surely gonna fail, I don't think there is point in continuing atleast in this case. Because that won't be comprehensive test output but useless verbose output on a sure thing that is always gonna happen.

---

For e.g. consider:

```bash
$ go test -v github.com/kedgeproject/kedge/pkg/spec -run TestParsePortMapping
=== RUN   TestParsePortMapping
=== RUN   TestParsePortMapping/Nothing_is_passed,_not_even_port
=== RUN   TestParsePortMapping/Only_'port'_is_passed
=== RUN   TestParsePortMapping/port:targetPort_is_passed
=== RUN   TestParsePortMapping/port/protocol_is_passed
=== RUN   TestParsePortMapping/port:targetPort/protocol_is_passed
=== RUN   TestParsePortMapping/Invalid_protocol_(neither_TCP_nor_UDP)_is_passed
=== RUN   TestParsePortMapping/Multiple_protocols_passed,_multiple_'/'_test
=== RUN   TestParsePortMapping/Non_int_port_is_passed
=== RUN   TestParsePortMapping/Non_int_targetPort_is_passed
=== RUN   TestParsePortMapping/More_than_2_ports_passed,_multiple_':'_test
--- FAIL: TestParsePortMapping (0.00s)
    --- PASS: TestParsePortMapping/Nothing_is_passed,_not_even_port (0.00s)
    --- PASS: TestParsePortMapping/Only_'port'_is_passed (0.00s)
    --- PASS: TestParsePortMapping/port:targetPort_is_passed (0.00s)
    --- PASS: TestParsePortMapping/port/protocol_is_passed (0.00s)
    --- FAIL: TestParsePortMapping/port:targetPort/protocol_is_passed (0.00s)
        resources_test.go:315: Expected test to pass but got an error -
                port is not an int: strconv.ParseInt: parsing "": invalid syntax
        resources_test.go:324: Expected ServicePort to be -
                <*>&ServicePort{Name:,Protocol:UDP,Port:1337,TargetPort:1338,NodePort:0,}
                But got -
                <nil>
    --- PASS: TestParsePortMapping/Invalid_protocol_(neither_TCP_nor_UDP)_is_passed (0.00s)
    --- PASS: TestParsePortMapping/Multiple_protocols_passed,_multiple_'/'_test (0.00s)
    --- PASS: TestParsePortMapping/Non_int_port_is_passed (0.00s)
    --- PASS: TestParsePortMapping/Non_int_targetPort_is_passed (0.00s)
    --- PASS: TestParsePortMapping/More_than_2_ports_passed,_multiple_':'_test (0.00s)
FAIL
exit status 1
FAIL    github.com/kedgeproject/kedge/pkg/spec  0.044s
```

So for failed test only this is sufficient

```bash
resources_test.go:315: Expected test to pass but got an error -
                port is not an int: strconv.ParseInt: parsing "": invalid syntax
```

We don't need to have the next set of lines:

```bash
        resources_test.go:324: Expected ServicePort to be -
                <*>&ServicePort{Name:,Protocol:UDP,Port:1337,TargetPort:1338,NodePort:0,}
                But got -
                <nil>
```

Reply:
A function like the following is being tested -
```golang
func a() {
  behavior x
  behavior y
  behavior z
}
```

We write a test -
```golang
func TestA() {
  t.Run("test a()", func(t *testing.T) {
    t.Fatal("fail if behavior x fails")
    t.Fatal("fail if behavior y fails")
    t.Fatal("fail if behavior z fails")
  }
}
```
Now, when behavior x fails, behavior y and behavior z will not be tested, in our test report we will only get failure for behavior x. When we fix that, we will then get an error for behavior y, then after fixing that we get the report for behavior z.

OTOH, if we use -

```golang
func TestA() {
  t.Run("test a()", func(t *testing.T) {
    t.Error("fail if behavior x fails")
    t.Error("fail if behavior y fails")
    t.Error("fail if behavior z fails")
  }
}
```
In this case we will get test failure reports for all the behaviors x, y and z at the same time and we can go ahead and do the fixes simultaneously.

The point of a test report is to find all the failures happening in the tests and making the report as extensive as we can so we can get what's breaking what, instead not to find failures in an iterative manner.